metadata/src/lib.rs

```diff
@@ -172,6 +174,31 @@ fn get_type_def_hash(
     hash(&bytes)
 }
 
+/// Check if the provided path segments construct a `node_template_runtime` type.
+///
+/// The `node_template_runtime::Call` contains dispatch calls to the pallets, registered
+/// in the same order as registered pallets.
+/// The presence of such structure needs to be taken into account when hashing:
+/// if the calls are simply hashed, then registering a different order of pallets
+/// would result in a different hashing **even if the pallets are firstly sorted
+/// in `get_metadata_per_pallet_hash` or `get_metadata_hash` functions **.
+fn check_template_runtime(segments: &[String]) -> bool {
+    let path_name = segments.join("::");
+
+    // The release version must match exactly the path segments.
+    if cfg!(not(test)) {
+        path_name == "node_template_runtime::Call"
+            || path_name == "node_template_runtime::Runtime"
+            || path_name == "node_template_runtime::Event"
+    } else {
+        // Testing cannot provide an absolute path of `node_template_runtime::Call`, but
+        // only a path ending in such segments.
+        path_name.ends_with("node_template_runtime::Call")
+            || path_name.ends_with("node_template_runtime::Runtime")
+            || path_name.ends_with("node_template_runtime::Event")
+    }
+}
```

Comment:
Changing logic based on whether we are testing or not is always best avoided if we can. We can construct our own `scale_info` types (instead of using `#[derive(TypeInfo)]`), which should allow you to manually create some types and set the paths to whatever you want. See:

https://docs.rs/scale-info/latest/scale_info/build/struct.TypeBuilder.html

I'd suggest having a play with that API in the tests before resroting to any conditional compilation like this :)